src/pushsource/_impl/backend/koji_source.py

```diff
@@ -283,18 +284,38 @@ def _push_items_from_rpm_meta(self, rpm, meta):
 
         candidate_paths = []
 
-        # If signing keys requested, try them in order of preference
-        for key in self._signing_key:
-            if key:
-                key = key.lower()
-                candidate = os.path.join(build_path, self._pathinfo.signed(meta, key))
-            else:
-                candidate = unsigned_path
-            candidate_paths.append(candidate)
-            if os.path.exists(candidate):
-                rpm_path = candidate
-                rpm_signing_key = key
+        timeout = int(os.getenv("PUSHSOURCE_SRC_POLL_TIMEOUT") or "0")
```

Comment:
It'd be preferable if there was less duplicated code between this and SourceWrapper.  For example, could there not be a function with signature:

```
def wait_exists(path: str):
    """Wait until the given path appears to exist, if enabled by environment variables.

    If the file does not exist after a certain timeout, logs a warning.
    """
```

Or if you had to, you could have:

```
def wait_any_exists(path: list[str]) -> Optional[str]:
   """Wait for any of the given paths to exist.

   Returns the first path confirmed to exist, or None if none of the paths
   exist by the deadline.
   """
```

Either one of which would be shareable between SourceWrapper and KojiSource rather than having a similar loop in both places.  But per later comments I doubt the second approach can be done safely anyway.

Reply:
In the current code iteration the logic seems different enough that it doesn't make sense to add it to the same function. Though the method has become quite long, so if you insist I'll try to think of a way to split the code.

Reply:
I don't think they are particularly different.  It seems to me like the SourceWrapper code could be:

```
for item in generator:
  if getattr(item, 'src', None):
    wait_exists(item.src)
  yield item
```

And the code in here could be:

```
if key:
    key = key.lower()
    candidate = os.path.join(build_path, self._pathinfo.signed(meta, key))
    wait_exists(candidate)
```

I suppose you could add an argument to wait_exists to customize the log message, but I'm not sure that's really necessary.

To me that code is much cleaner.  I'm not going to insist on it though, but can you please not put me in that position ("I will improve the code only if you insist on it")?  If you think my suggestion is not an improvement, just say so.